src/drivers/imx/dummy-dma.c

```diff
@@ -45,54 +160,358 @@ static int dummy_dma_status(struct dma_chan_data *channel,
+
+	if (curr_idx >= channel->desc_count)
+		return false;
+
+	next->elem = ch->elems->elems[curr_idx];
+	return true;
+}
+
+/**
+ * \brief Perform one individual copy
+ * \param[in] ch: DMA channel to do the copy for
+ * \param[in, out] next: Next element
+ * \return: Number of bytes copied, if more data can be copied.
+ *          Number of bytes copied negated, if the copy must stop short.
+ *
+ * Does a copy of one elem. The elem itself is either from the config or set
+ * by a IRQ callback using DMA_CB_STATUS_SPLIT.
+ *
+ * Updates @next within the callback. If the callback set next.status as
+ * DMA_CB_STATUS_RELOAD, @next is additionally updated to get the next elem
+ * from the config.
+ */
+static int do_onecopy(struct dma_chan_data *ch, struct dma_cb_data *next)
+{
+	struct dma_chan_pdata *pdata = dma_chan_get_data(ch);
+	/* Perform a single elem copy */
+	void *dest = (void *)next->elem.dest;
+	void *src = (void *)next->elem.src;
+	size_t size = next->elem.size;
+
+	/* Do the copy */
+	/* Note that we don't know the real buffer size,
+	 * we assume `size` is correct for memcpy_s
+	 */
+	memcpy_s(dest, size, src, size);
+
+	/* For status, point to immediately after the last copy */
+	pdata->r_pos = next->elem.src + size;
+	pdata->w_pos = next->elem.dest + size;
+
+	if (!(ch->cb_type & DMA_CB_TYPE_IRQ))
+		return -size; /* Callback not applicable, end for now */
+
+	/* Call the registered DMA IRQ callback to finish processing of this
+	 * elem. Default to loading the next descriptor.
+	 */
+	next->status = DMA_CB_STATUS_RELOAD;
+
+	ch->cb(ch->cb_data, DMA_CB_TYPE_IRQ, next);
+
+	/* Interpret what the callback set in next */
+	switch (next->status) {
+	case DMA_CB_STATUS_RELOAD:
+		/* Load next elem from the config */
+		return get_next_elem(ch, next) ? size : -size;
+	case DMA_CB_STATUS_SPLIT:
+		/* We got new elem in next; we may just return */
+		return size;
+	case DMA_CB_STATUS_END:
+	case DMA_CB_STATUS_IGNORE:
+		/* This copy should end */
+		return -size;
+	default:
+		trace_dummydma_error("dummy_dma docopy: switch default case");
+		trace_dummydma_error("dummy_dma docopy: status is %d",
+				     next->status);
+		/* Stop now */
+		return -size;
+	}
+}
+
+/**
+ * \brief Perform the actual copying.
+ * \param[in] ch: Channel to perform the copying on
+ * \param[in] bytes: Minimum number of bytes to be copied
+ * \return The number of bytes actually copied
+ *
+ * Copy @bytes bytes on channel @ch.
+ * The copying may stop short if a callback returned DMA_CB_TYPE_IGNORE,
+ * DMA_CB_TYPE_END or if the elems have been exhausted from the config.
+ */
+static int do_copy(struct dma_chan_data *ch, int bytes)
+{
+	/* Allocate a static "next" variable */
+	/* TODO Maybe save it in the channel data?? */
+	struct dma_cb_data next;
+	int orig_bytes = bytes;
+
+	/* Just don't do any copy if there is no way to configure it */
+	if (!get_next_elem(ch, &next))
+		return 0;
+
+	for (;;) {
+		int size = do_onecopy(ch, &next);
+
+		if (size <= 0) {
```

Comment:
Move declaration of size at the beginning of the function. No need for new line after call to do_onecopy